### DISCUSSION TITLE
About StorageProvider VERSION

### DISCUSSION THREAD

**Discussion by strong84** [openmls/openmls#1677] [Q&A] [answered]
When implementing the [StorageProvider](https://book.openmls.tech/traits/traits.html#storageprovider) trait. I have some questions about the VERSION.

According to the Book, it says: 
`The traits are generic over a VERSION, which is used to ensure that the values that are persisted can be upgraded when OpenMLS changes the stored structs.`

According to my understanding, I need to save this VERSION to database. so later on, if any object structure changes, for example, `QueuedProposal` or `KeyPackage`. the VERSION can tell if the serialized bytes is from the previous generation of the Object. 

1. I see that all those objects are using `serde_json` to serialize/deserialize in MemoryStorage. Does it provide backward compatibility? What if I use `tls_codec`?

2. The VERSION looks like user-defined. I can provide any number. Is this to say: It is application's responsibility to manage the migration? or the `VERSION` has to match with the `CURRENT_VERSION` in  StorageProvider trait?

 Wish to know more about the specifics. Thanks~

**Comment by kkohbrok** [comment, marked as answer]
Thanks for the question and apologies for the late response.

Your understanding is correct in that the purpose of VERSION is to allow OpenMLS to change data structures and to signal these changes on the type level. This is to ensure that OpenMLS won't take old storage provider implementations as input that have not yet been updated to work with the new or updated structs.

MemoryStorage is mostly for debugging purposes. Data stored in MemoryStorage is not expected to be persisted anyway and thus won't have to be migrated.

Managing the VERSION is OpenMLS' responsibility. If changes in OpenMLS require existing structs to change or for new structs to be stored, it will change the CURRENT_VERSION to indicate that. It's then up to the storage provider implementer to migrate the respective storage (although OpenMLS will probably provide some guidance or even some code to migrate the affected structs). We haven't done that yet, so for now it's all theoretical.

Does that answer your questions?

**Reply by strong84** [reply]
Sounds like the migration process has not yet finalized. Thanks for the answer! 👍

**Reply by kkohbrok** [reply]
The migration mechanism is finalized and storage provider implementers should (in one way or another) include VERSION when persisting the individual pieces of data. We just haven't had to increment the CURRENT_VERSION yet, so so far it's not been used outside of a test case.
